### ISSUE TITLE
Mounting file as volume inconsistently reflects file changes

### ISSUE THREAD

**Issue by whitecolor** [docker/for-win#1392]
Seems to be the same issue as https://github.com/docker/for-win/issues/328 

Docker version 17.09.0-ce, build afdb6d4

Mounting host file as volume.

### Expected behavior

File changes on the host to be consistently reflected in a container.

### Actual behavior

File changes on the host to be **inconsistently** reflected in a container.

### Steps to reproduce the behavior
1. Create a folder with simple `file.txt` in folder for example `C:/_docker_mount_test`, place some text there, for example: add line `test`.
2. Run test container `docker run -it --rm -v C:/_docker_mount_test/file.txt:/file.txt -v C:/_docker_mount_test:/folder alpine:3.1 sh`
3. check file content in container: `cat /file.txt` (should be `test`)
4. change file content on the host (for example add one more line `test2`)
5. check file content in container: `cat /file.txt` - you probably should see something different (probably just `test`)
6. check file content in mounted folder `cat /foler/file.txt`, it should be correct.
7. if you check `cat /file.txt` content after this, it should be correct.
![image](https://user-images.githubusercontent.com/736697/33708655-dab18306-db5c-11e7-978f-fb62969bd059.png)

![image](https://user-images.githubusercontent.com/736697/33708902-92d0f142-db5d-11e7-9f87-6d518c972732.png)

There is maybe and explanation given in https://github.com/moby/moby/issues/15793#issuecomment-135411504 
```
If you are using some editor like vim, when you save the file it does not save the file directly, rather it creates a new file and copies it into place.
This breaks the bind-mount, which is based on inode. Since saving the file effectively changes the inode, changes will not propagate into the container.
```
But I tried just to update the file with `echo something >> file.txt`, and the issue is still here.

I wonder why docker shared volumes mechanics can not handle such case?

**Comment by joaociocca** [comment]
I just did a crash course on Docker and started using it with https://github.com/HiromuHota/pentaho-kettle. When I first save a file inside the container, it first shows up nicely on the host... but any edit won't reflect. If I delete the file and try again, nothing happens.

**Comment by whitecolor** [comment]
This seems to be "expected" behavior when mounting a single file, though it is still buggy.

**Comment by zhangysh1995** [comment]
New folders on my host are not displayed inside the docker container.

Inside:
````
test@13a477cfe79a:~/data$ ls
Arduino  cpu_features  git    netdata  obs-studio  shadowsocks-libev
CMake    curl          kitty  nuklear  php-src     tmux
FFmpeg   domoticz      lab    numpy    radare2     vim
````

On my host:(** means missing folders)
````
[zhangyushan@sbtest2 data]$ ls
**accelerInt**      cpu_features  kitty        nuklear     radare2
Arduino         curl          lab          numpy       shadowsocks-libev
**cantera**         FFmpeg        **lighttpd1.4**  obs-studio  tmux
**contrail-build**  git           netdata      php-src     vim
````

Docker version 17.09.0-ce, build afdb6d4
CentOS Linux release 7.3.1611 (Core)

**Comment by pscheit** [comment]
this is still an actual bug i think. If I update the bound file in the container the changes are reflected.

It is not an issue of the editor used (nano from WSL behaves the same as phpstorm without safe write enabled).

**Comment by pscheit** [comment]
/remove-lifecycle stale
I dont think this is solved yet? I still have to mount directories instead of files everywhere

**Comment by pengisgood** [comment]
I got the same issue.
```
docker version
Client:
 Version:           18.06.1-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        e68fc7a
 Built:             Tue Aug 21 17:24:56 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.1-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       e68fc7a
  Built:            Tue Aug 21 17:23:21 2018
  OS/Arch:          linux/amd64
  Experimental:     true
```

**Comment by bradmccormack** [comment]
I'm seeing this too

/remove-lifecycle stale

**Comment by ruslantalpa** [comment]
+1

**Comment by cheffe** [comment]
Currently in a training workshop. We all have the issue. We are running a nginx instance and try to alter a local index.html. The first change is refreshed and served, the second change onwards is not.

**Comment by whitecolor** [comment]
@cheffe, for now, I would recommend just attach folders with files

**Comment by cheffe** [comment]
Thanks, using a folder instead of a file works, even on Windows.

**Comment by MaciejKucia** [comment]
This is a duplicate of https://github.com/moby/moby/issues/15793

**Comment by maresb** [comment]
@MaciejKucia although this issue presents the same as moby/moby#15793 the underlying problem is completely different.  The issue in moby/moby#15793 was platform-independent, and was caused by "safe-mode writing" of files (i.e. instead of directly writing to a file, you make a new copy and rename it).  It's more of a problem with particular text editors than with Docker.

In contrast, this issue appears to be a legitimate bug in Docker for Windows.  Docker for Windows can mount directories but it simply fails when a file is mounted and then updated.

Interestingly, I have noticed that although the contents of a file remain the same, the last-modified time gets updated.  Bizarre.
